File: model/src/main/java/io/github/runtimemodels/chazm/event/RoleEventFactory.java
package io.github.runtimemodels.chazm.event;

import io.github.runtimemodels.chazm.entity.Role;

/**
 * The {@linkplain RoleEventFactory} interface defines the API for constructing {@linkplain RoleEvent}s.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
@FunctionalInterface
public interface RoleEventFactory {

    /**
     * Constructs an {@linkplain RoleEvent}.
     *
     * @param category the {@linkplain EventCategory}.
     * @param role     the {@linkplain Role}.
     * @return a {@linkplain RoleEvent}.
     */
    RoleEvent build(EventCategory category, Role role);

}

File: chazm-model/src/main/kotlin/runtimemodels/chazm/model/entity/RoleEventFactory.kt
package runtimemodels.chazm.model.entity

import runtimemodels.chazm.api.entity.Role
import runtimemodels.chazm.model.event.EventType

/**
 * The [RoleEventFactory] interface defines the API for constructing [RoleEvent]s.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
@FunctionalInterface
interface RoleEventFactory {
    /**
     * Constructs an [RoleEvent].
     *
     * @param category the [EventType].
     * @param role     the [Role].
     * @return a [RoleEvent].
     */
    fun build(category: EventType, role: Role): RoleEvent
}
